### PR TITLE
swarm-derive: Specify `NetworkBehaviour` by full path

### PR DESCRIPTION
**Pull request by folex**
# Description

Specify `NetworkBehaviour` as `::libp2p::swarm::NetworkBehaviour` in derive.

## Links to any relevant issues

https://github.com/libp2p/rust-libp2p/issues/2921

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by mxinden** [comment]
@folex can you grant me permissions to merge `master` into your branch? I have a bunch of pull requests lined up, thus having me do it will safe a bit of time on both ends.

File: swarm-derive/src/lib.rs [hunk 1]
```diff
@@ -125,7 +125,7 @@ fn build_struct(ast: &DeriveInput, data_struct: &DataStruct) -> TokenStream {
                             )
                             .unwrap();
                             let ty = &field.ty;
-                            quote! {#variant(<#ty as NetworkBehaviour>::OutEvent)}
+                            quote! {#variant(<#ty as ::libp2p::swarm::NetworkBehaviour>::OutEvent)}
```
Comment (mxinden) [inline swarm-derive/src/lib.rs]:
```suggestion
                            quote! {#variant(<#ty as #trait_to_impl>::OutEvent)}
```

How about reusing `trait_to_impl`?

**Review by mxinden — COMMENTED** [review]
Thanks for preparing the patch. I will make sure it makes it into `v0.49.0` (https://github.com/libp2p/rust-libp2p/pull/2931).

File: swarm-derive/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 - Fix an issue where the derive would generate bad code if the type parameters between the behaviour and a custom
   out event differed. See [PR 2907].
+- Fix an issue where the derive would generate incorrect code depending on available imports [#2921]
 
 [PR 2907]: https://github.com/libp2p/rust-libp2p/pull/2907
```
Comment (thomaseizinger) [inline swarm-derive/CHANGELOG.md]:
```suggestion
- Fix an issue where the derive would generate incorrect code depending on available imports. See [PR 2921].

[PR 2907]: https://github.com/libp2p/rust-libp2p/pull/2907
[PR 2921]: https://github.com/libp2p/rust-libp2p/pull/2921
```

**Review by thomaseizinger — APPROVED** [review]
Thanks for sending this patch.

LGTM modulo the changelog and @mxinden's suggestion!

**Comment by folex** [comment]
> @folex can you grant me permissions to merge `master` into your branch? I have a bunch of pull requests lined up, thus having me do it will safe a bit of time on both ends.

I don't see how to do that in GitHub without actually adding you to the fluencelabs org -_-

**Comment by thomaseizinger** [comment]
> > @folex can you grant me permissions to merge `master` into your branch? I have a bunch of pull requests lined up, thus having me do it will safe a bit of time on both ends.
> 
> I don't see how to do that in GitHub without actually adding you to the fluencelabs org -_-

There is a checkbox next to the people that are subscribed to the PR that will allow us to push to this branch only.

**Comment by folex** [comment]
> > > @folex can you grant me permissions to merge `master` into your branch? I have a bunch of pull requests lined up, thus having me do it will safe a bit of time on both ends.
> > 
> > 
> > I don't see how to do that in GitHub without actually adding you to the fluencelabs org -_-
> 
> There is a checkbox next to the people that are subscribed to the PR that will allow us to push to this branch only.

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Seems like that doesn't work with organisations, only with personal forks. 🤔

**Pull request merged**